BranchMetrics/android-branch-deep-linking-attribution
# [minor][SDK-1592][SDK-1832] Support for deferred init for plugins

## Reference
SDK-1592 -- [Android] Native changes to support onPluginReady callback
SDK-1832 -- Android - update branch.json fields and support

## Description

### Scenario 
Integrations using our plugin SDKs (React-Native, Capacitor, Unity, etc), it is possible to have a race condition wherein the native layers finish their initialization before the JS/C# layers have finished loading and registering their receivers- dropping the Branch parameters.
     
Because these plugin delays are not deterministic, or consistent, a constant offset to delay is not guaranteed to work in all cases, and possibly penalizes performant devices.

To solve, we wait for the plugin to signal when it is ready*, and then begin native init. To do so, we expose a new method for plugins and save the session object as it was created, only if configured through the `branch.json`.

*This logic is purely up to the plugins but generally it would be safe to `notify` once the listeners have been registered plugin-side.

### Why this is not default behavior
Not all apps have this problem, and this solution optimizes for correctness. The downside, however, is that the network request is made only after it is told to do so by the plugin, meaning what was once parallel (and sometimes too early) is now sequential (but always on time). 

## Testing Instructions
Add `branch.json` with the follow contents to your assets folder:

```
{
    "liveKey": "key",
    "testKey": "key_test",
    "enableLogging": true,
    "deferInitForPluginRuntime": true
}
```

Changes inside the plugin to use this feature should be invisible to the end user. Tentatively in our React-Native plugin, the the public `subscribe` method will check if we need to notify the native layer. https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/commit/c785782836a9bf0ece5f289873816325e168f5d9#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R47

Additionally, a button to manually notify the SDK to init has been added to the test bed to simulate the flow. Queue up a link and some events to verify that the SDK does not automatically init, and all options/params are preserved.

## Risk Assessment [`MEDIUM`]
The new functionality is purely opt in and would require intent to configure. However, it is possible that there could be some misconfiguration that could lead to the SDK not initializing properly. The logs will clearly state if the new feature is responsible.

- [✅ ] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- Correctness & Style

cc @BranchMetrics/saas-sdk-devs for visibility.
